File: computer/actions/scan.lua
local network = require "network"
local location = require "device.location"

---@param body ScanRequest
return function(body)
    local scanner = peripheral.find("geoScanner")
    if scanner == nil then
        return
    end

    local position = location.getPosition()
    local blocks = scanner.scan(body.range)

    local data = {}
    for index, block in ipairs(blocks) do
        local pos = {
            x = block.x,
            y = block.y,
            z = block.z
        }
        data[#data + 1] = {
            name = block.name,
            offset = pos
        }
    end

    network.broadcastPacket({
        type = "response:scan",
        range = body.range,
        blocks = data,
        position = position
    })
end

File: server/lib/packet/actions/scan.ts
import { z } from "zod";

import { Vec3 } from "../generic";

export const ScanRequest = z.object({
    type: z.literal("request:scan"),
    range: z.number(),
});
export type ScanRequest = z.infer<typeof ScanRequest>;

export const ScanResponse = z.object({
    type: z.literal("response:scan"),
    position: Vec3.optional(),
    range: z.number(),
    blocks: z
        .object({
            name: z.string(),
            offset: Vec3,
        })
        .array(),
});
export type ScanResponse = z.infer<typeof ScanResponse>;
